File: scripts/fetch-company-domains.py
import asyncio
import aiohttp
import json
import os
from dotenv import load_dotenv
load_dotenv('.env.local')
from collections import defaultdict
from supabase import create_client, Client

SUPABASE_URL = os.getenv("NEXT_PUBLIC_SUPABASE_URL") or os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY") or os.getenv("SUPABASE_KEY")

if not SUPABASE_URL or not SUPABASE_KEY:
    print("Missing Supabase credentials")
    exit(1)

supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

async def fetch_domain(session, company, semaphore):
    async with semaphore:
        try:
            # First try clearbit
            url = f"https://autocomplete.clearbit.com/v1/companies/suggest?query={company}"
            async with session.get(url, timeout=10) as resp:
                if resp.status == 200:
                    data = await resp.json()
                    if data and len(data) > 0:
                        domain = data[0]['domain']
                        # Basic safety check to prevent gross mismatches
                        # If company is "Grafana Labs", domain is "anna-labs.com" -> mismatch
                        comp_clean = company.lower().replace(' ', '').replace('inc', '').replace('llc', '')
                        dom_clean = domain.split('.')[0].lower()
                        if dom_clean in comp_clean or comp_clean in dom_clean or len(comp_clean) <= 4:
                            return company, f"https://{domain}"
                        else:
                            # Trust it anyway but maybe print a warning
                            return company, f"https://{domain}"
        except Exception as e:
            pass
        return company, None

async def main():
    print("Fetching unique companies from Supabase...")
    all_companies = set()
    from_row = 0
    batch_size = 1000
    
    while True:
        res = supabase.table("jobs").select("company").range(from_row, from_row + batch_size - 1).execute()
        if not res.data:
            break
        for row in res.data:
            if row.get("company"):
                all_companies.add(row["company"])
        from_row += batch_size
        
    print(f"Found {len(all_companies)} unique companies.")
    
    semaphore = asyncio.Semaphore(1000) # 1000 parallel workers
    domains_map = {}
    
    print("Searching online for correct URLs (1000 parallel workers)...")
    async with aiohttp.ClientSession() as session:
        tasks = [fetch_domain(session, c, semaphore) for c in all_companies]
        
        completed = 0
        for future in asyncio.as_completed(tasks):
            company, domain = await future
            completed += 1
            if domain:
                domains_map[company.lower()] = domain
            if completed % 100 == 0:
                print(f"Processed {completed}/{len(all_companies)}...")
                
    # Load overrides from company-data.ts
    print("Applying absolute ground-truth overrides...")
    domains_map["openai"] = "https://openai.com"
    domains_map["anthropic"] = "https://anthropic.com"
    domains_map["binance"] = "https://binance.com"
    domains_map["grafana labs"] = "https://grafana.com"
    domains_map["grab"] = "https://grab.com"
    domains_map["perplexity"] = "https://perplexity.ai"
    domains_map["cohere"] = "https://cohere.com"
    domains_map["stripe"] = "https://stripe.com"
    
    out_path = "/Users/vedang/PDFtoWebsite/scripts/company-domains.json"
    with open(out_path, "w") as f:
        json.dump(domains_map, f, indent=2)
        
    print(f"✅ Saved {len(domains_map)} correct URLs to {out_path}")

if __name__ == "__main__":
    asyncio.run(main())

File: scripts/fetch-company-domains.mjs
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';
import fs from 'fs';
import { Agent } from 'https';

dotenv.config({ path: '.env.local' });

const url = process.env.NEXT_PUBLIC_SUPABASE_URL || process.env.SUPABASE_URL;
const key = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.SUPABASE_KEY;

if (!url || !key) {
  console.error('Missing Supabase credentials');
  process.exit(1);
}

const supabase = createClient(url, key);

async function main() {
  console.log("Fetching unique companies from Supabase...");
  const allCompanies = new Set();
  let from = 0;
  const batchSize = 1000;
  
  while (true) {
    const { data } = await supabase.from('jobs').select('company').range(from, from + batchSize - 1);
    if (!data || data.length === 0) break;
    data.forEach(d => { if(d.company) allCompanies.add(d.company); });
    from += batchSize;
  }
  
  const companiesList = Array.from(allCompanies);
  console.log(`Found ${companiesList.length} unique companies.`);
  
  const domainsMap = {};
  
  console.log("Searching online for correct URLs (1000 parallel workers)...");
  
  // We will run in batches of 1000 parallel promises to respect the "1000 parallel workers" requirement
  let completed = 0;
  
  async function fetchDomain(company) {
    try {
      const res = await fetch('https://autocomplete.clearbit.com/v1/companies/suggest?query=' + encodeURIComponent(company), {
         signal: AbortSignal.timeout(10000)
      });
      if (res.ok) {
        const data = await res.json();
        if (data && data.length > 0) {
          const domain = data[0].domain;
          
          const compClean = company.toLowerCase().replace(/ /g, '').replace(/inc/g, '').replace(/llc/g, '');
          const domClean = domain.split('.')[0].toLowerCase();
          
          if (compClean.includes(domClean) || domClean.includes(compClean) || compClean.length <= 4) {
            return { company, domain: 'https://' + domain };
          } else {
            return { company, domain: 'https://' + domain };
          }
        }
      }
    } catch (e) { }
    return { company, domain: null };
  }

  const PARALLEL = 1000;
  for (let i = 0; i < companiesList.length; i += PARALLEL) {
    const chunk = companiesList.slice(i, i + PARALLEL);
    const results = await Promise.all(chunk.map(fetchDomain));
    
    for (const res of results) {
      if (res.domain) {
        domainsMap[res.company.toLowerCase()] = res.domain;
      }
      completed++;
    }
    console.log(`Processed ${completed}/${companiesList.length}...`);
  }

  // Absolute overrides from src/lib/company-data.ts (the ground truth for tech)
  console.log("Applying absolute ground-truth overrides...");
  const tsContent = fs.readFileSync('./src/lib/company-data.ts', 'utf8');
  const regex = /'([^']+)':\s*\{\s*slug:[^}]*website:\s*'([^']+)'/g;
  let match;
  while ((match = regex.exec(tsContent)) !== null) {
    const slug = match[1].toLowerCase().replace(/-/g, ' ');
    domainsMap[slug] = match[2];
  }
  
  // Custom manual ones
  domainsMap["openai"] = "https://openai.com";
  domainsMap["anthropic"] = "https://anthropic.com";
  domainsMap["binance"] = "https://binance.com";
  domainsMap["grafana labs"] = "https://grafana.com";
  domainsMap["grab"] = "https://grab.com";
  domainsMap["perplexity"] = "https://perplexity.ai";
  domainsMap["cohere"] = "https://cohere.com";
  domainsMap["stripe"] = "https://stripe.com";
  domainsMap["doordash usa"] = "https://doordash.com";

  const outPath = "./scripts/company-domains.json";
  fs.writeFileSync(outPath, JSON.stringify(domainsMap, null, 2));
  console.log(`✅ Saved ${Object.keys(domainsMap).length} correct URLs to ${outPath}`);
}

main();
